### ISSUE TITLE
Karma + Node.js 17: can't connect to karma server. There is no server listening on port 9876

### ISSUE THREAD

**Issue by marcoscaceres** [karma-runner/karma#3730]
When running under Node 17, running karma reports:
```
 can't connect to karma server. There is no server listening on port 9876 
```

Switching back to Node 16 seems to "fix" the issue.  

See also:
https://youtrack.jetbrains.com/issue/WEB-53632

**Comment by marcoscaceres** [comment]
In case it helps, here are some dependencies (in case it's one of those that is causing it):

```
    "karma": "^6.3.9",
    "karma-chrome-launcher": "^3.1.0",
    "karma-firefox-launcher": "^2.1.2",
    "karma-jasmine": "^4.0.1",
    "karma-jasmine-html-reporter": "^1.7.0",
    "karma-mocha-reporter": "^2.2.5",
    "karma-safari-launcher": "^1.0.0",
```

**Comment by marcoscaceres** [comment]
Some steps to reproduce.

1. nvm install 17; nvm use 17; 
2. git clone git@github.com:w3c/respec.git
3. npm i
4. npm start -- --browser Chrome --grep "W3C"

**Comment by devoto13** [comment]
I've tried the above steps using Node v17.1.0, Chrome 96.0.4664.55 on macOS 12.0.1 and the tests ran successfully. I've also tried our own test suite which also completed successfully. I've tried to run using Karma configuration from WebStorm and it worked as well. Can you double check on your side?

**Comment by devoto13** [comment]
Reproduced on CI: https://github.com/karma-runner/karma/runs/4304771965?check_suite_focus=true. I'll look more into it tomorrow.

**Comment by marcoscaceres** [comment]
Thanks! and yeah, we were able to reproduce it on a different Mac (cc @sideshowbarker) running different versions of MacOS (Monterey and Big Sur)... so, at least, it doesn't seem to be an OS thing, but more likely a Node 17 thing.

**Comment by ChaitanyaBabar** [comment]
+1

**Comment by eweitz** [comment]
This issue also occurs with Node 18, unfortunately.

**Comment by devoto13** [comment]
https://github.com/nodejs/node/issues/40702 seems to be the culprit. Node 17 change the DNS resolution, so now it resolves `localhost` according to the OS settings instead of IPv4 first. The Karma server only listens on IPv4 address (`0.0.0.0`) by default, but the requests are sent to `localhost` and `localhost` is resolved into IPv6 address (`::`) in Node 17+. So the run request is unable to reach the Karma server. 

I'll think what's the best way to make it work out of the box. In the meantime you can address the issue by setting [`listenAddress`](http://karma-runner.github.io/6.3/config/configuration-file.html#listenaddress) to `::` either in config or using `LISTEN_ADDR` environment variable.

**Comment by treysis** [comment]
Why not make listening on `::` the default instead of forcing IPv4?

**Comment by devoto13** [comment]
Because some environments may not support IPv6, that would be a breaking change for them. Primarily thinking about Docker on Windows/Mac here, but there could be others. I've added an item https://github.com/karma-runner/karma/issues/3503 to consider this for the next major release.

**Comment by cherniavskii** [comment]
Hey @devoto13 
Did you consider omitting `host` parameter in `server.listen()`?
According to the docs, IPv6/IPv4 are handled automatically:
> If host is omitted, the server will accept connections on the [unspecified IPv6 address](https://en.wikipedia.org/wiki/IPv6_address#Unspecified_address) (::) when IPv6 is available, or the [unspecified IPv4 address](https://en.wikipedia.org/wiki/0.0.0.0) (0.0.0.0) otherwise.
Source: https://nodejs.org/docs/latest-v16.x/api/net.html#serverlistenport-host-backlog-callback

Context: I've found this thread while investigating why karma server doesn't start on my setup (macOS 12.4, node v16.15.0) with default `listenAddress`.
I'm setting `listenAddress` to `::` or `localhost` for now as a workaround, but I wonder if the suggestion above would solve the issue once and for all?

**Comment by cherniavskii** [comment]
Hey @devoto13 
Thanks for looking into this.

Indeed, I just realised that the root cause of the issue is that I use VPN (NordVPN to be specific).
Without it, the server starts just fine, so I guess VPN is messing with my network configuration.

Anyway, I've changed karma `net-utils.js` code locally to omit `host` - and it works both with VPN and without it.
For now I can use `listenAddress: 'localhost'` as a workaround.
